A few fixes which might need discussion

frank-dittrich opened the pull request:
1. report glibc version with --list=build-info
2. Add _BSD_SOURCE and _DEFAULT_SOURCE definitions to listconf.c for HAVE_OPENCL (We use setenv() here, but didn't have the feature request definitions...)
3. Fix some memory leaks for ./john --status
4. Report build timestamp with --list=build-info
5. Report MS compiler version with --list=build-info

magnumripper commented on src/listconf.c:
```diff
@@ -63,6 +67,8 @@ extern void cuda_device_list();
 #if HAVE_OPENCL
 #include "common-opencl.h"
+#define _BSD_SOURCE 1
+#define _DEFAULT_SOURCE 1
```
Feature macros are supposed to be defined prior to including *any* system header so these should be moved up above the `#define NEED_OS_FLOCK`.

magnumripper commented:
I'm not overly excited about recovery.c; It's an error condition, no need to bother. But that's just me. The rest is fine (once you move the feature macros).

frank-dittrich commented:
@magnum I moved the _DEFAULT_SOURCE and _BSD_SOURCE definition up, and I added two more commits.
The build time stamp might be helpful to let users know how old their john version is (and that it might be a good idea to get a newer version).
I had no way of testing the "Add Microsoft compiler version info..." commit.
Who can test this, now that Jim is inactive?

magnumripper commented:
On Yosemite using a real gcc:
```
listconf.c:57:30: fatal error: gnu/libc-version.h: No such file or directory
 #include <gnu/libc-version.h>
                              ^
compilation terminated.
make[1]: *** [listconf.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [default] Error 2
```
I'll #ifdef that code away for now.

magnumripper commented:
Temporary band-aid in 3b54b18. Using __GNUC__ was not a good idea, I should have noticed that.

frank-dittrich commented:
Sorry for that.
Meanwhile, I also reverted the second part of that commit, to avoid a --list=build-info segfault under Linux.
